src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -890,6 +892,103 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
     return parts_with_ranges;
 }
 
+void MergeTreeDataSelectExecutor::filterPartsByQueryConditionCache(
+    RangesInDataParts & parts_with_ranges,
+    const SelectQueryInfo & select_query_info,
+    const ContextPtr & context,
+    LoggerPtr log)
+{
+    const auto & settings = context->getSettingsRef();
+    if (!settings[Setting::use_query_condition_cache] || !settings[Setting::allow_experimental_analyzer] ||
+        (!select_query_info.prewhere_info && !select_query_info.filter_actions_dag))
+        return;
+
+    QueryConditionCachePtr query_condition_cache = context->getQueryConditionCache();
+
+    struct Stat
+    {
+        size_t total_granules{0};
+        size_t granules_dropped{0};
+    };
+
+    auto drop_mark_ranges =[&](const ActionsDAG::Node * dag) -> Stat
+    {
+        size_t condition_hash = dag->getHash();
+        Stat stat;
+        for (auto it = parts_with_ranges.begin(); it != parts_with_ranges.end();)
+        {
+            auto & part_with_ranges = *it;
+            stat.total_granules += part_with_ranges.getMarksCount();
+
+            auto & data_part = part_with_ranges.data_part;
+            auto storage_id = data_part->storage.getStorageID();
+            auto matching_marks_opt = query_condition_cache->read(storage_id.uuid, data_part->name, condition_hash);
+            if (!matching_marks_opt)
+            {
+                ++it;
+                continue;
+            }
+
+            auto & matching_marks = *matching_marks_opt;
+            MarkRanges ranges;
+            for (auto & mark_range : part_with_ranges.ranges)
+            {
+                size_t begin = mark_range.begin;
+                for (size_t mark_it = begin; mark_it < mark_range.end; ++mark_it)
+                {
+                    if (!matching_marks[mark_it])
+                    {
+                        ++stat.granules_dropped;
+                        if (mark_it == begin)
+                            ++begin;
+                        else
+                        {
+                            ranges.emplace_back(begin, mark_it);
+                            begin = mark_it + 1;
+                        }
+                    }
+                }
+
+                if (begin != mark_range.begin && begin != mark_range.end)
+                    ranges.emplace_back(begin, mark_range.end);
+                else if (begin == mark_range.begin)
+                    ranges.emplace_back(begin, mark_range.end);
+            }
+
+            if (ranges.empty())
+                it = parts_with_ranges.erase(it);
+            else
+            {
+                part_with_ranges.ranges = ranges;
+                ++it;
+            }
+        }
+
+        return stat;
+    };
+
+    if (const auto & prewhere_info = select_query_info.prewhere_info)
+    {
+        for (const auto * dag : prewhere_info->prewhere_actions.getOutputs())
```

Comment:
What about [PrewhereInfo::row_level_filter](https://github.com/ClickHouse/ClickHouse/blob/a85679444656c8163cc61a90824cdac8f9c2299e/src/Storages/SelectQueryInfo.h#L46)? It does not look ok to ignore it. Some marks might have been skipped because of it and the following queries that have the same prewhere condition will skip those marks even if the row policy is no longer there.

I just ran into this problem in one of my PRs: https://github.com/ClickHouse/ClickHouse/pull/85118. The [test file](https://github.com/ClickHouse/ClickHouse/pull/85118/files#diff-353a9eb71e0756bac1b98f19bf0f6bab84e177b8b9a45bbf8b324453cb60d9adR1) can be used as a repro by removing the `cache=0`. Bear in mind this repro works only with a build produced by that PR